software/teachee-desktop/src/app.rs

```diff
@@ -82,29 +112,143 @@ impl eframe::App for App {
+                        ui.columns(2, |uis| {
+                            uis[0].label("Offset");
+                            uis[0].add(Slider::new(v_offset, 0.0..=100.0).show_value(false));
+                            uis[1].label("Scale");
+                            uis[1].add(Slider::new(v_scale, 0.0..=100.0).show_value(false));
+                        });
+
+                        ui.add_space(GROUP_SPACING);
+                        ui.separator();
+                        ui.add_space(GROUP_SPACING);
+
+                        ui.vertical_centered_justified(|ui| {
+                            if ui
+                                .add(Button::new("Reset").min_size((0.0, BUTTON_HEIGHT).into()))
+                                .clicked()
+                            {
+                                *h_offset = 0.0;
+                                *h_scale = 0.0;
+                                *v_offset = 0.0;
+                                *v_scale = 0.0;
+                                *saved_v_offset = 0.0;
+                                *saved_v_scale = 0.0;
+                                *channel1_v_offset = 0.0;
+                                *channel1_v_scale = 0.0;
+                                *channel2_v_offset = 0.0;
+                                *channel2_v_scale = 0.0;
+                                *channel1_on = false;
+                                *channel2_on = false;
+                            }
+                        });
+
+                        ui.add_space(GROUP_SPACING);
+                        // TODO: replace with new controls
+                        channel_control(ui, "Channel 1", channel1, channel1_offset);
+                        channel_control(ui, "Channel 2", channel2, channel2_offset);
+                    });
+
+                    ui.add_space(GROUP_SPACING);
+
+                    ui.group(|ui| {
+                        ui.label("Triggers");
+                        ui.add_space(GROUP_SPACING);
+                        ui.separator();
+                        ui.add_space(GROUP_SPACING);
+                        ui.vertical_centered_justified(|ui| {
+                            if ui
+                                .add(
+                                    Button::new(trigger_button_text.to_string())
+                                        .min_size((0.0, BUTTON_HEIGHT).into()),
+                                )
+                                .clicked()
+                            {
+                                use TriggerControl::*;
+                                // TODO: enable/disable triggering
+                                *trigger_button_text = match trigger_button_text {
+                                    Start => Stop,
+                                    Stop => Start,
+                                };
+                            }
+                        });
+                    });
+                });
+            });
+
```

Comment:
I cannot comment much on the UI code but I like the way it looks. One extension that I think would be really handy is to also have a text box available for entering the offsets and scales rather than sliders. For instance, if you are trying to scope a signal at a known frequency, it would be really helpful to just type "1" s/div or V/div into a text box to set the scale. The sliders should also be kept though as they replicate a physical knob you see on other scopes.

Reply:
What would the units for offset be? Do we need to manually enter offsets? I think scale is enough based on other oscilloscopes.

Reply:
I will add editable text boxes that display the s/div and channel 1 and 2 V/div and perhaps others